src/entry.c

```diff
@@ -0,0 +1,349 @@
+ *
+ * Entry with value pointer, used for larger fields and values. The field is SDS
+ * type 8 or higher.
+ *
+ *     +-------+--------------+
+ *     | value | field        |
+ *     | ptr   | hdr "foo" \0 |
+ *     +-------+------^-------+
+ *                    |
+ *                    |
+ *                 entry pointer = field sds
+ */
+
+/* The maximum allocation size we want to use for entries with embedded
+ * values. */
+#define EMBED_VALUE_MAX_ALLOC_SIZE 128
+
+/* SDS aux flag. If set, it indicates that the entry has TTL metadata set. */
+#define FIELD_SDS_AUX_BIT_ENTRY_HAS_TTL 0
+
+/* SDS aux flag. If set, it indicates that the entry has an embedded value
+ * pointer located in memory before the embedded field. If unset, the entry
+ * instead has an embedded value located after the embedded field. */
+#define FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR 2
+
+bool entryHasValuePtr(const entry *entry) {
+    return sdsGetAuxBit(entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR);
+}
+
+bool entryHasExpiry(const entry *entry) {
+    return sdsGetAuxBit(entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_TTL);
+}
+
+/* Returns the location of a pointer to a separately allocated value. Only for
+ * an entry without an embedded value. */
+sds *entryGetValueRef(const entry *entry) {
+    serverAssert(entryHasValuePtr(entry));
+    char *field_data = sdsAllocPtr(entry);
+    field_data -= sizeof(sds *);
+    return (sds *)field_data;
+}
+
+/* The entry pointer is the field sds, but that's an implementation detail. */
+sds entryGetField(const entry *entry) {
+    return (sds)entry;
+}
+
+sds entryGetValue(const entry *entry) {
+    if (entryHasValuePtr(entry)) {
+        return *entryGetValueRef(entry);
+    } else {
+        /* Skip field content, field null terminator and value sds8 hdr. */
+        size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
+        serverAssert((char *)entry + offset);
+
+        return (char *)entry + offset;
+    }
+}
+
+entry *entrySetValue(entry *e, sds value) {
+    if (entryHasValuePtr(e)) {
+        sds *value_ref = entryGetValueRef(e);
+        sdsfree(*value_ref);
+        *value_ref = value;
+        return e;
+    } else {
+        entry *new_entry = entryUpdate(e, value, entryGetExpiry(e));
+        return new_entry;
+    }
+}
+
+/* Returns the address of the entry allocation. */
+void *entryAllocPtr(const entry *entry) {
+    char *buf = sdsAllocPtr(entry);
+    if (entryHasValuePtr(entry)) buf -= sizeof(sds *);
+    if (entryHasExpiry(entry)) buf -= sizeof(long long);
+    return buf;
+}
+
+/**************************************** Entry Expiry API *****************************************/
+long long entryGetExpiry(const entry *entry) {
+    long long expiry = EXPIRY_NONE;
+    if (entryHasExpiry(entry)) {
+        char *buf = sdsAllocPtr(entry);
+        if (entryHasValuePtr(entry)) buf -= sizeof(sds *);
+        buf -= sizeof(expiry);
+        memcpy(&expiry, buf, sizeof(expiry));
```

Comment:
Memcpy is fine, but it's aligned so it should also be safe to access it as a long-long-pointer?

```suggestion
        expiry = *(long long *)buf;
```

Reply:
Yes. It can be done. initially there was also an optional metadata before the expiration which is why I used memcpy. but you are right we can change it now